sklearn/ensemble/_gb.py

```diff
@@ -636,6 +645,11 @@ def _fit_stages(
                     raw_predictions[~sample_mask],
                     sample_weight[~sample_mask],
                 )
+                if i == 0:
+                    self.oob_scores_[i] = old_oob_score
+                else:
+                    self.oob_scores_[i] = old_oob_score - self.oob_improvement_[i]
+                self.oob_oob_score_ = self.oob_scores_[-1]
```

Comment:
I think that we need more refactoring than just introducing the attribute:

```python
            if do_oob:
                sample_mask = _random_sample_mask(n_samples, n_inbag, random_state)
                if i == 0:
                    initial_loss = loss_(
                        y[~sample_mask],
                        raw_predictions[~sample_mask],
                        sample_weight[~sample_mask],
                    )

            # fit next stage of trees
            raw_predictions = self._fit_stage(
                i,
                X,
                y,
                raw_predictions,
                sample_weight,
                sample_mask,
                random_state,
                X_csc,
                X_csr,
            )

            # track deviance (= loss)
            if do_oob:
                self.train_score_[i] = loss_(
                    y[sample_mask],
                    raw_predictions[sample_mask],
                    sample_weight[sample_mask],
                )
                self.oob_scores_[i] = loss_(
                    y[~sample_mask],
                    raw_predictions[~sample_mask],
                    sample_weight[~sample_mask],
                )
                previous_loss = initial_loss if i == 0 else self.oob_scores_[i - 1]
                self.oob_improvement_[i] = previous_loss - self.oob_scores_[i]
                self.oob_oob_score_ = self.oob_scores_[-1]
```

We can compute the `initial_loss` only for `i == 0` and not afterwards since it is already stored in `oob_scores_[i - 1]`. Then, we can deduce the improvement from the `oob_scores_` directly.